Applications search & pagination docs (#1898)
FusionAuth/fusionauth-site: site/docs/v1/tech/apis/applications.adoc

include::docs/v1/tech/apis/_application-response-body.adoc[]

== Search for Applications

[NOTE.since]
====
This API has been available since 1.44.0
====

This API is used to search for Applications and may be called using the `GET` or `POST` HTTP methods. Examples of each are provided below. The `POST` method is provided to allow for a richer request object without worrying about exceeding the maximum length of a URL. Calling this API with either the `GET` or `POST` HTTP method will provide the same search results given the same query parameters.

=== Request

[.api-authentication]
link:/docs/v1/tech/apis/authentication#api-key-authentication[icon:lock[type=fas]] Search for Applications using the given search and sort criteria
[.endpoint]
.URI
--
[method]#GET# [uri]#/api/application/search``?name=\{name\}&active=\{active\}&tenantId=\{tenantId\}&orderBy={field}&startRow={startRow}``#
--

==== Request Parameters

[field]#name# [type]#[String]# [optional]#Optional#::
The string to search for in the Application name. This can contain wildcards using the asterisk or percent characters (* or %).

[field]#active# [type]#[Boolean]# [optional]#Optional#::

> Filters on the Application's `active` status.

Reviewer: I would mention that omitting this parameter or specifying `null` will retrieve applications in both statuses.

> Filters on the Application's `active` status.

Reviewer: I would mention that omitting this parameter or specifying `null` will retrieve applications in both statuses.

[field]#tenantId# [type]#[UUID]# [optional]#Optional#::

> Restricts the results to Applications belonging to the given Tenant.

Reviewer: Mention that specifying the tenant Id through a header value or scoped API key will override the value on the request. Same for `POST`.

[field]#numberOfResults# [type]#[Integer]# [optional]#Optional# [default]#defaults to `25`#::
The number of results to return from the search.

[field]#orderBy# [type]#[String]# [optional]#Optional# [default]#defaults to `insertInstant DESC`#::
The database column to order the search results on plus the order direction.
+
The columns you can use for this are:
+
* `id` - the UUID of the Application
* `insertInstant` - the link:/docs/v1/tech/reference/data-types#instants[instant] when the Application was created
* `name` - the Application name

> * `tenant` - the id of the Tenant in which the Application belongs

Reviewer: This actually sorts by the tenant name.

+

> For example, to order the results by the insert instant in a descending order, the value would be provided as `insertInstant DESC`. The final string is optional can be set to `ASC` or `DESC`.

Reviewer: Mention that if `ASC` and `DESC` are omitted, it will default to `ASC`.

I would assume most users know that leaving it out in a SQL query will sort in ascending order, but it could be different here.

Reviewer: I see now that this comes from another search API. It's probably good.

[field]#startRow# [type]#[Integer]# [optional]#Optional# [default]#defaults to `0`#::
The offset row to return results from. If the search has 200 records in it and this is 50, it starts with row 50.

> 

Reviewer: After the request parameters, include file for tenant header blurb. Same for `POST` operation.
```
include::docs/v1/tech/apis/_x-fusionauth-tenant-id-header-scoped-operation.adoc[]
```

[.api-authentication]
link:/docs/v1/tech/apis/authentication#api-key-authentication[icon:lock[type=fas]] Search for Applications using the given search and sort criteria
[.endpoint]
.URI
--
[method]#POST# [uri]#/api/application/search#
--

When calling the API using a `POST` request you will send the search criteria in a JSON request body.

==== Request Body

[.api]
[field]#search.name# [type]#[String]# [optional]#Optional#::

> The string to search for in the Application name. This can contain wildcards using the asterisk or percent characters (* or %).

Reviewer: Is it worth saying that it is a case insensitive search? Or do you think that's implied? (same applies to `GET`)

[field]#search.active# [type]#[Boolean]# [optional]#Optional#::

> Filters on the Application's `active` status.

Reviewer: Mention `null` here as well.

[field]#search.tenantId# [type]#[UUID]# [optional]#Optional#::
Filters Applications to those belonging to the specified Tenant.

[field]#search.numberOfResults# [type]#[Integer]# [optional]#Optional# [default]#defaults to `25`#::
The number of results to return from the search.

[field]#search.orderBy# [type]#[String]# [optional]#Optional# [default]#defaults to `insertInstant DESC`#::
The database column to order the search results on plus the order direction.
+
The columns you can use for this are:
+
* `id` - the UUID of the Application
* `insertInstant` - the link:/docs/v1/tech/reference/data-types#instants[instant] when the Application was created
* `name` - the Application name

> * `tenant` - the id of the Tenant in which the Application belongs

Reviewer: Name of tenant.

+

> For example, to order the results by the insert instant in a descending order, the value would be provided as `insertInstant DESC`. The final string is optional can be set to `ASC` or `DESC`.

Reviewer: I believe you'll find this is indented with the bulleted list. Add a space before each of the `*` in the list to fix it. Example from event log search:
<img width="845" alt="image" src="https://user-images.githubusercontent.com/3409780/216455898-fb2b6480-3d83-4c0d-bac1-886242731189.png">

Reviewer: You could update it over there too if you want :wink:

include::docs/v1/tech/apis/_application-response-body.adoc[]

== Search for Applications

[NOTE.since]
====
This API has been available since 1.44.0
====

This API is used to search for Applications and may be called using the `GET` or `POST` HTTP methods. Examples of each are provided below. The `POST` method is provided to allow for a richer request object without worrying about exceeding the maximum length of a URL. Calling this API with either the `GET` or `POST` HTTP method will provide the same search results given the same query parameters.

=== Request

> 

Reviewer 2: ```suggestion
[.api-authentication]
link:/docs/v1/tech/apis/authentication#api-key-authentication[icon:lock[type=fas]] Search for Applications by name
[.endpoint]
.URI
--
[method]#GET# [uri]#/api/application/search``?name=\{name\}``#
--

[.api-authentication]
link:/docs/v1/tech/apis/authentication#api-key-authentication[icon:lock[type=fas]] Search for Applications by tenant
[.endpoint]
.URI
--
[method]#GET# [uri]#/api/application/search``?tenantId=\{tenantId\}``#
--

```

These seem like the most common two options, so it is worth making them explicit.

Author: 👍 Should we also illustrate sorting and paging on one of these? If so, do we provide a concrete example, like `orderBy=name`, or keep it generic like `orderBy={field}`?

Reviewer 2: you are already (below) illustrating sorting. I don't think we need to illustrate paging, we do that here: https://fusionauth.io/docs/v1/tech/core-concepts/search#pagination

Though you could change the verbiage from 'users and entities' to 'search results'. (And also add a section here that talks about searches for other things, but that would be a nice to have.)

Reviewer 2: @mmanes sorry if I was unclear. I thought we could have three sample urls

* one with all the parameters (which you had before my comment)
* one with application name
* one with tenant

[.api-authentication]
link:/docs/v1/tech/apis/authentication#api-key-authentication[icon:lock[type=fas]] Search for Applications using the given search and sort criteria
[.endpoint]
.URI
--
[method]#GET# [uri]#/api/application/search``?name=\{name\}&active=\{active\}&tenantId=\{tenantId\}&orderBy={orderBy}&startRow={startRow}``#
--

> ==== Request Parameters

Reviewer 3: We usually alphabetize parameters.

[field]#name# [type]#[String]# [optional]#Optional#::

> The string to search for in the Application name. This can contain wildcards using the asterisk or percent characters (* or %). This search parameter is case-insensitive.

Reviewer 2: ```suggestion
The string to search for in the Application name. This can contain wildcards using the asterisk character (`*`). This parameter is case-insensitive.
```

> The string to search for in the Application name. This can contain wildcards using the asterisk or percent characters (* or %). This search parameter is case-insensitive.

Reviewer 2: I guess we do support both wildcards for audit logs, but I think it is needlessly complex.

> The string to search for in the Application name. This can contain wildcards using the asterisk or percent characters (* or %). This search parameter is case-insensitive.

Reviewer 3: I tend to agree with this. Probably simpler to just document one option. 

To note...  if you omit the `*` we will do a "fuzzy" search - for example, if you search for `foo` it will be converted to `%foo%`. But if you search for `foo*` it will be converted to `foo%`, so using a wild card in this example is more specific than omitting it. 

This is just to say, we don't currently have a way to ask a search API to do an exact match like we do in Elasticsearch.

> The string to search for in the Application name. This can contain wildcards using the asterisk or percent characters (* or %). This search parameter is case-insensitive.

Author: @mooreds @robotdan How does this sound?
```The case-insensitive string to search for in the Application name. This can contain wildcards using the asterisk character (*). If no wildcards are present, the search criteria will be interpreted as "*value*". If this field is omitted, no filtering will be applied to the name field.```

> The string to search for in the Application name. This can contain wildcards using the asterisk or percent characters (* or %). This search parameter is case-insensitive.

Reviewer 3: > If this field is omitted, no filtering will be applied to the name field.

Not sure if we have to say this - if we do, we'd need to add it to all search criteria fields, and I think it is implied.

> [field]#active# [type]#[Boolean]# [optional]#Optional#::

Reviewer 2: What is the default here?

Reviewer 2: Looks like it is to retrieve all applications. Maybe put a note to see the body for the default, like we do for this field: `application.samlv2Configuration.logout.keyId`.

Author: I feel that omitting a search field implies that we're not going to filter on that field, which is different than our other "defaults" where we actually force in a value. We could add a `Defaults to [see description]` blurb, but I feel that it adds noise. 

Anyway, just my $.02. Happy to something here if you'd like.

Reviewer 3: I agree - this is a case where an optional does not have a default value. W/out any search criteria, you end up with just a paginated list of applications using a documented order by.

> Filters on the Application's `active` status. Can be omitted or set to `null` to retrieve both active and inactive Applications.

Reviewer 3: We officially don't support using `null` in our APIs.

Reviewer: That was my bad.

[field]#tenantId# [type]#[UUID]# [optional]#Optional#::
Restricts the results to Applications belonging to the given Tenant. This parameter will be overridden if the request contains an `X-FusionAuth-TenantId` header, or if the supplied API key is scoped to a specific Tenant.

[field]#numberOfResults# [type]#[Integer]# [optional]#Optional# [default]#defaults to `25`#::
The number of results to return from the search.

> [field]#orderBy# [type]#[String]# [optional]#Optional# [default]#defaults to `insertInstant DESC`#::

Reviewer 3: More for the code review, but we have a mix of default sort orders, some are `insertInstant ASC` and some are `insertInstant DESC`, I wonder if we should standardize. 

Also, when sorting on `insertInstant` as a default, should we add a secondary sort order? Such as `insertInstant DESC, name ASC` ?

Author: This also happened to be a mistake. The default sort for this is `name ASC`.

It's possible that multiple rows could be inserted in the same millisecond, and without a secondary sort we could miss rows around the page boundaries.

> The database column to order the search results on plus the order direction.

Reviewer 2: ```suggestion
The database field to order the search results as well as an order suffix.
```

+

> The columns you can use for this are:

Reviewer 2: ```suggestion
The fields you can use for this are:
```

I don't know we want to reveal this is a database column, and we don't use that term in the user search API.

> The columns you can use for this are:

Reviewer 3: Sort of a minor issue, but we use the text `The possible values are:` in a lot of places to document the available values for a parameter. 

I do see the Audit Log Search (maybe others) are using the text you have:`The columns you can use for this are:`

I am ok with either, but we should do a quick audit to ensure they are all the same.

+
 * `id` - the UUID of the Application
 * `insertInstant` - the link:/docs/v1/tech/reference/data-types#instants[instant] when the Application was created
 * `name` - the Application name

>  * `tenant` - the name of the Tenant in which the Application belongs

Reviewer 3: Isn't this `tenantId`?

Reviewer: The search parameter is `tenantId`. The sort field is `tenant`.

+

> For example, to order the results by the insert instant in a descending order, the value would be provided as `insertInstant DESC`. The final string is optional can be set to `ASC` or `DESC`. If omitted, the default sort order is `ASC`.

Reviewer 2: What happens if I put something other than `ASC` or `DESC`?

Reviewer 2: ```suggestion
The order suffix  is optional. Possible values of the order suffix are `ASC` or `DESC`. If omitted, the default sort order is `ASC`.
+
For example, to order the results by the insert instant in a descending order, use `insertInstant DESC`.
```

Reviewer 2: I moved this around because I wanted to explain each part of the string before showing an example.

Also, I don't love the term `order suffix` because it feels fancy, but couldn't think of a better term.

Reviewer: Maybe `order direction`? I think that most developers will be familiar.

I just confirmed that specifying something other than `ASC` or `DESC` results in a validation error.

```
The [search.orderBy] property contains an invalid column name. The invalid column name is [insertInstant GAR]. The supported values are [id, name, insertInstant].
```

Does that validation message need to be expanded to mention valid sort directions? That would be a change to existing APIs as well. I'm not sure whether changing a validation error message is considered a breaking change.

Reviewer 2: I think order direction is fine.

I don't know about whether modifying an API error message is a breaking change. I think `no` because we don't document them anywhere, but am not sure.

However, that error message isn't great, because it doesn't tell you that `GAR` is the issue. I'll leave it up to you whether you want to change it.

[field]#startRow# [type]#[Integer]# [optional]#Optional# [default]#defaults to `0`#::

> The offset row to return results from. If the search has 200 records in it and this is 50, it starts with row 50.

Reviewer 3: It may be worth doing a quick audit to ensure we are consistely describing this parameter across all of our search APIs. 

Here is the text from the Entity Types and Login Records APIs.

> The offset into the total results. In order to paginate the results, increment this value by the [field]#numberOfResults# for subsequent requests.

While it is sort of obvious - I kind of like adding the `increment this value by the [field]#numberOfResults# for subsequent requests.` to help developers know how to paginate.

Author: Combining them both may be a good option, like: 
```The offset row to return results from. If the search has 200 records in it and this is 50, it starts with row 50. In order to paginate the results, increment this value by the [field]#numberOfResults# for subsequent requests.```

Reviewer 3: I think we have a mix.. but in general it doesn't read very well to end a sentence with `from`. 

Here is how we say it other places in the API docs which I think reads better, and may be more grammatically correct. But I know we have some grammar nerds.. so I'll defer there. 

> The offset into the total results.

&nbsp;

For the second part of the description, this sounds like an example, so I would prefer to use our "For example..." syntax. Here is a suggestion:

```
[field]#startRow# [type]#[Integer]# [optional]#Optional# [default]#defaults to `0`#::
The offset into the total results. In order to paginate the results, increment this value by the [field]#numberOfResults# for subsequent requests.
+
For example, if the total search results are greater than the page size designated by [field]#numberOfResults#,  to request the second page of results, assuming the default page size, set this value to `25` to retrieve results `26-50`.
```

include::docs/v1/tech/apis/_x-fusionauth-tenant-id-header-scoped-operation.adoc[]

[.api-authentication]

> link:/docs/v1/tech/apis/authentication#api-key-authentication[icon:lock[type=fas]] Search for Applications using the given search and sort criteria

Reviewer 3: Same comments from the `GET` version

Author: I _believe_ your concerns were addressed here the changes refactored into `_application-search-request-parameters.adoc`.

[.endpoint]
.URI
--
[method]#POST# [uri]#/api/application/search#
--

> When calling the API using a `POST` request you will send the search criteria in a JSON request body.

Reviewer 2: Bigger picture question, why is support for `POST` needed? For the user search, people can create complex queries that might exceed the url limit. But I don't see that happening here.

Reviewer: Mainly because we are following the existing patterns and extending `BaseAPISearchAction` which allows for both `GET` and `POST`. Mark already doesn't like that we have a `POST` method to query data.

Reviewer 2: okay.

==== Request Body

[.api]
[field]#search.name# [type]#[String]# [optional]#Optional#::

> The string to search for in the Application name. This can contain wildcards using the asterisk or percent characters (* or %). This search parameter is case-insensitive.

Reviewer 2: Same comment about multiple wildcards

> [field]#search.active# [type]#[Boolean]# [optional]#Optional#::

Reviewer 2: same comment about default

Filters on the Application's `active` status. Can be omitted or set to `null` to retrieve both active and inactive Applications.

[field]#search.tenantId# [type]#[UUID]# [optional]#Optional#::
Filters Applications to those belonging to the specified Tenant. This parameter will be overridden if the request contains an `X-FusionAuth-TenantId` header, or if the supplied API key is scoped to a specific Tenant.

[field]#search.numberOfResults# [type]#[Integer]# [optional]#Optional# [default]#defaults to `25`#::
The number of results to return from the search.

[field]#search.orderBy# [type]#[String]# [optional]#Optional# [default]#defaults to `insertInstant DESC`#::

> The database column to order the search results on plus the order direction.

Reviewer 2: Same comment about reworking this, order suffix, etc. In fact, this is large enough that I would suggest moving it to an include file so that you can share the text between the GET and the POST sections. I think you could probably do everything like that. That makes understanding the doc more complex, but prevents future maintainers from adding a parameter or modifying a desc to the `GET` section and not the `POST` section.

Here's an example: https://raw.githubusercontent.com/FusionAuth/fusionauth-site/master/site/docs/v1/tech/apis/_user-search-request-parameters.adoc

Reviewer: Perfect! Thank you for the example. Mark and I talked about this. I think we will use this pattern to repeat the request parameters for `GET` and `POST`.

+
The columns you can use for this are:
+
 * `id` - the UUID of the Application
 * `insertInstant` - the link:/docs/v1/tech/reference/data-types#instants[instant] when the Application was created
 * `name` - the Application name
 * `tenant` - the name of the Tenant in which the Application belongs
+
For example, to order the results by the insert instant in a descending order, the value would be provided as `insertInstant DESC`. The final string is optional can be set to `ASC` or `DESC`. If omitted, the default sort order is `ASC`.

> [field]#startRow# [type]#[Integer]# [optional]#Optional# [default]#defaults to `0`#::

Reviewer 3: Should this be `search.startRow` ?

Reviewer 3: This should be `search.startRow`.

Author: This code was outdated. The field prefix has been fixed in the refactored `_application-search-request-parameters.adoc`.

The offset row to return results from. If the search has 200 records in it and this is 50, it starts with row 50.

[source,json]
.Example JSON Request
----
include::docs/src/json/applications/search-post-request.json[]
----

> include::docs/v1/tech/apis/_x-fusionauth-tenant-id-header-scoped-operation.adoc[]

Reviewer 2: I don't know if you need the API tenant include, but if you do, it should be above the JSON request.

Reviewer: _Just_ above the JSON request?

Reviewer 2: hmmm. Good question. I don't think we're 100% consistent. The headers should be above the request body, but below the request parameters.

So:

urls
request parameters (if any)
request headers
request body 
example request

Reviewer: How should we structure this for search APIs where we have both GET and POST? Would it be like this?

GET URLs
GET request parameters (query string)
GET request headers

POST URLs
POST request headers
POST request body
Example POST request

Reviewer 2: That is correct. We should put this in the DevDocsReadme, too. Can you do that, @spwitt , please? 

```
Request section header
GET URLs (could have 1-3 of these, show the most common)
GET request parameters (query string)
GET request headers

POST URLs (only will be one, typically)
POST request headers
POST request body
Example POST request(s)

Response section header
Response codes
Response body
Example response(s)
```

=== Response

The response for this API contains the Applications matching the search criteria in paginated format.

> include::docs/v1/tech/apis/_standard-get-response-codes.adoc[]

Reviewer 2: I don't think you'll ever get a 503, will you? Since this is against the database?

Looks like that's in the audit log search body as well, but again, not sure that is possible.

You can set `:never_search_error:` and then unset it to omit the `503`.

Author: Done. I also added `:never_missing:` since we don't ever return a `404` here.

==== Response Body

[.api]

> [field]#applications# [type]#[Array]#::

Reviewer 3: I don't think we have any other APIs that don't document the object returning in the results.

Author: The object is documented earlier on the page, should we repeat it, or simply reference/link to it?

Author: I've went ahead and added it, using the include used earlier in the page.

Reviewer 3: Yes, it should be added here, even if it is duplicated.

The list of Applications returned by the search.

[field]#total# [type]#[Integer]#::

> The total number of Applications matching the search criteria. Use this value along with the [field]#numberOfResults# and [field]#startRow# in the Search request to perform pagination.

Reviewer 2: Why is `Search` capitalized.

Reviewer: All of the other instances of this message capitalize it (Group members, Event logs, Audit logs).

Should we not?

Reviewer 2: Ah, then, for sure, leave it capitalized. We have a bit of an issue with capitalizing various entities, but this isn't the time to fix it.
